### PR TITLE
cql3: convert LWT IF clause to expressions

### PR DESCRIPTION
**Pull request by avikivity**
LWT `IF` (column_condition) duplicates the expression prepare and evaluation code. Annoyingly,
LWT IF semantics are a little different than the rest of CQL: a NULL equals NULL, whereas usually
NULL = NULL evaluates to NULL.

This series converts `IF` prepare and evaluate to use the standard expression code. We employ
expression rewriting to adjust for the slightly different semantics.

In a few places, we adjust LWT semantics to harmonize them with the rest of CQL. These are pointed
out in their own separate patches so the changes don't get lost in the flood.

### CODE REVIEW

**Comment by avikivity** [comment]
@cvybhu This is almost working (need to fix up the commits). Only problem (so far) is `IF some_column IN (1, 2, NULL)` since lists reject NULL.

I thought of several options, all look bad.

Here are the two least-worst:
1. create a new type list_with_nulls that does accept nulls (lots of work)
2. allow lists to accept nulls, only check when creating mutations out of them (risk: we'll miss a spot, and end up with NULLs in storage)
3. allow frozen lists to accept NULLs (in storage too), I think we can easily catch non-frozen lists during conversion to atomic cells

**Comment by avikivity** [comment]
eh, that's three

**Comment by avikivity** [comment]
I fixed up the commits, except the last two, which are still problematic.

**Comment by nyh** [comment]
Issue #5825 about improper validation of nulls in list might be related.

**Comment by avikivity** [comment]
> Issue #5825 about improper validation of nulls in list might be related.

Absolutely, I'll try to sanitize this area (before proceeding with the series).

**Comment by avikivity** [comment]
> > Issue #5825 about improper validation of nulls in list might be related.
> 
> Absolutely, I'll try to sanitize this area (before proceeding with the series).

Turns out that's a driver bug.

**Comment by avikivity** [comment]
Rebased on top of #12411 . Still some work to do (pass expression directly from parser).

/cc @gusev-p

**Comment by avikivity** [comment]
Checkpoint: rebased past https://github.com/scylladb/scylladb/pull/12411

**Comment by avikivity** [comment]
- rebased (thanks @cvybhu for binary_operator prepare_
- relaxed some LWT constraints/oddities in preparation
- tests now pass. 

The first six commits are part of https://github.com/scylladb/scylladb/pull/12411, so don't review them here.

**Comment by avikivity** [comment]
Update:
 - drop empty patch (rebase artifact)
 - dropped comment about changed tests in next-to-last patch, those problems were identified and extracted to preceding patches

**Comment by avikivity** [comment]
- rebased past merged 94331081585daffcf144cf2938d6d8bf1b42c150

**Comment by nyh** [comment]
Unfortunately I got confused and reviewed https://github.com/scylladb/scylladb/pull/12597 instead of this one :-( So after we finish the discussion there, I can merge this PR...

**Comment by gusev-p** [comment]
> @cvybhu @gusev-p review ping

I'm looking into it.

File: cql3/expr/prepare_expr.cc
Comment (cvybhu) [inline cql3/expr/prepare_expr.cc:1335]: I wonder if we also check types that contain duration, like `list<duration>`.
Maybe they fail at runtime?

File: cql3/expr/expression.cc
Comment (cvybhu) [inline cql3/expr/expression.cc:209]: Printing in error messages should use non-debug printing using `expression::printer`

File: cql3/column_condition.cc
Comment (cvybhu) [inline cql3/column_condition.cc:81]: We could get the kespace and table name for `column_specification` constructor from the `schema`.

File: cql3/expr/expression.cc
Comment (avikivity) [reply cql3/expr/expression.cc:1589]: It's better to use `=` and `!=` to avoid surprising the users, even if we lost a little bit of information here.

File: cql3/column_condition.cc
Comment (avikivity) [reply cql3/column_condition.cc:76]: They're indeed not allowed.

File: cql3/expr/expression.cc
Comment (gusev-p) [inline cql3/expr/expression.cc:290]: I would prefer to have something like
```
enum class null_handling_style {
   standard,
   special
};
compare(managed_bytes_opt, managed_bytes_opt, null_handling_style) -> optional<bool_or_null>;
```
and not copy-paste `equal`, `not_equal`, `is_one_of_special_null`

File: cql3/expr/prepare_expr.cc
Comment (gusev-p) [inline cql3/expr/prepare_expr.cc:1320]: By returning non-null we are not going inside the expr-tree, just check the top of it. Is it really what we want here?

File: cql3/selection/selection.cc
Comment (gusev-p) [inline cql3/selection/selection.cc:160]: `cdefs.reserve`?

File: cql3/expr/expression.cc
Comment (nyh) [reply cql3/expr/expression.cc:1589]: I agree. If the users used "=" in an expression, they would want to see "=" in the error message, not "EQUALS_SPECIAL_NULL" or "=*".
By the way, now that I think of it, I have some doubts if we really need the EQ_SPECIAL_NULL operator. Instead of being a different operator, it could have been an option to evaluate() on how to evaluate the regual EQ operator.

Comment (cvybhu) [reply cql3/expr/expression.cc:1589]: I agree that a flag in `evaluation_inputs` that specifies in which context we are evaluating would be cleaner
I think the problem is that we would have to remember to set this flag in all the right places. With this solution we just generate the right data in the parser and then we don't have to worry that we might have missed setting the flag somewhere.
Another option would be having the flag inside `binary_operator`, something like `is_special_lwt_null_binop`

File: cql3/expr/expression.cc
Comment (avikivity) [reply cql3/expr/expression.cc:209]: I filed #12702 because I refuse to use expression::printer.

File: cql3/expr/expression.cc
Comment (avikivity) [reply cql3/expr/expression.cc:257]: It was unused before, so I'll leave it. Eventually we'll just drop evaluate_binop_sides, in fact it does very little already.

File: cql3/expr/expression.cc
Comment (avikivity) [reply cql3/expr/expression.cc:539]: Fixed outside the series.

File: cql3/expr/expression.cc
Comment (avikivity) [reply cql3/expr/expression.cc:289]: Removed.

File: cql3/expr/expression.cc
Comment (avikivity) [reply cql3/expr/expression.cc:290]: I'm thinking we should move the null handling style into binary_operator, instead of adding new enums. It will make not changing </<=/>/>= meaning easier. Next revision will have it so.

File: cql3/expr/expression.cc
Comment (avikivity) [reply cql3/expr/expression.cc:1589]: Changed to have a separate enum in binary_operator.

Printing now shows plain `=` in user mode, but annotated in debug mode.

File: cql3/expr/expression.cc
Comment (avikivity) [reply cql3/expr/expression.cc:2087]: It's required indirectly, since we convert `x LIKE constant` to `func(x)`.

Functions handle NULL already, it's just this path that was broken (and isn't used in selection).

File: cql3/expr/prepare_expr.cc
Comment (avikivity) [reply cql3/expr/prepare_expr.cc:1320]: search_and_replace recurses.

Comment (avikivity) [reply cql3/expr/prepare_expr.cc:1320]: ah, but I broke the recursion. Will fix.

File: cql3/expr/expression.cc
Comment (avikivity) [reply cql3/expr/expression.cc:386]: I changed `is_one_of()` (which now contains `is_one_of_special_null()`) to return `FALSE` for this case, to be more consistent with LWT, which doesn't return NULL (though it's rather invisible, since the condition is consumed internally and not exposed).

File: cql3/expr/expression.cc
Comment (avikivity) [reply cql3/expr/expression.cc:553]: Fixed.

File: cql3/expr/expression.cc
Comment (avikivity) [reply cql3/expr/expression.cc:560]: Fixed.

File: cql3/expr/expression.cc
Comment (avikivity) [reply cql3/expr/expression.cc:1601]: It's not `IN` in user mode and annotated in debug mode.

File: cql3/expr/prepare_expr.cc
Comment (avikivity) [reply cql3/expr/prepare_expr.cc:1335]: references_duration() recurses into collection types and checks the element types.

File: cql3/selection/selection.cc
Comment (avikivity) [reply cql3/selection/selection.cc:160]: Added.

File: cql3/expr/expression.cc
Comment (avikivity) [reply cql3/expr/expression.cc:209]: Fixed (using {:user}).

**Comment by avikivity** [comment]
Update:


 - dropped `static` from function in anonymous namespace                                                                                                                                                                             
 - changed from SPECIAL_NULL_{EQ,NEQ,IN} to orthogonal enum                                                                                                                                                                          
 - restored < > <= >= LWT NULL semantics (throw if second operand is NULL)                                                                                                                                                           
   - fewer tests changed                                                                                                                                                                                                             
 - printout is now unchanged in user mode                                                                                                                                                                                            
 - is_one_of(x, NULL) now returns FALSE for LWT null handling (instead of                                                                                                                                                            
   NULL); doesn't affect anything since the result is consumed internally and                                                                                                                                                        
   not exposed to the user                                                                                                                                                                                                           
 - use list_contains_null() in one more place rather than open-coding it                                                                                                                                                             
 - call to evaluate(..., evaluation_inputs{}) (not forwarding real inputs)                                                                                                                                                           
   fixed (by eliminating the call site, is_one_of_special_null(), in favor of                                                                                                                                                        
   is_one_of())                                                                                                                                                                                                                      
 - fixed optimize_like() when LIKE is not a top-level expression; added unit                                                                                                                                                         
   test as penance                                                                                                                                                                                                                   
 - split removal of oper_is_slice() from restrictions.cc into its own patch,                                                                                                                                                         
   and justify the removal                                                                                                                                                                                                           
 - added vector::reserve() to selection_from_partition_slice()                                                                                                                                                                       
 - error message citing expression when subscripting non-map/non-list uses                                                                                                                                                           
   user-oriented expression printing                                                                                                                                                                                                 
 - disambiguated documentation of map element access with NULL subscript                                                                                                                                                             
 - adjusted cql-pytest to keep passing on Cassandra

File: cql3/column_condition.cc
Comment (nyh) [inline cql3/column_condition.cc]: I know this has already been decided, but I want to point out again that all these error cases like `list[-1]` being an error, were useful for catching buggy applications. Now `list[-1]` evaluates to null it will be harder for people who write applications to find bugs in the application.

File: test/cql/lwt_test.result
Comment (nyh) [inline test/cql/lwt_test.result:2052]: Nitpick: to check if c[-1] is null, maybe it's better to check "c[-1] is null" instead of c[-1] = false?
(but I don't know, is IS NOT implemented in LWT? maybe it's not implemented).

File: cql3/column_condition.cc
Comment (nyh) [inline cql3/column_condition.cc:22]: Here too the mention of EQ_SPECIAL_NULL in the commit message is no longer correct.

File: cql3/column_condition.cc
Comment (nyh) [inline cql3/column_condition.cc:55]: If I understand correctly, these three calls copy _expr three times, even if there is no need (e.g., optimize_like() will usually not do anything). Could these functions be changed to modify the expression in-place instead of always creating a new copy?

**Review by nyh — APPROVED** [review]
Looks good (and impressive).
There are two outdated commit messages refering to EQ_SPECIAL_NULL that you might want to fix.
I also left a bunch of minor nitpicks you may want to consider and possibly reject.

File: cql3/expr/expression.cc
Comment (avikivity) [reply cql3/expr/expression.cc:267]: Will fix.

File: cql3/expr/expression.cc
Comment (avikivity) [reply cql3/expr/expression.cc:346]: Absolutely. The previous version eliminated the difference in favor of allowing NULL everywhere, but a reviewer objected and I restored the old behavior.

I plan to add `USING DIALECT OPTION = 'SQL' | 'LWT'` so people can move to SQL syntax and leave this quagmire beind.

File: cql3/expr/prepare_expr.cc
Comment (avikivity) [reply cql3/expr/prepare_expr.cc:1247]: I could change it to `like(<the original pattern>)`.

File: cql3/expr/prepare_expr.cc
Comment (avikivity) [reply cql3/expr/prepare_expr.cc:1290]: It's part of function's interface. I propose not to change function's interface in this already long series.

File: cql3/expr/expression.cc
Comment (avikivity) [reply cql3/expr/expression.cc:2660]: A topic for another series.

File: cql3/column_condition.cc
Comment (avikivity) [reply cql3/column_condition.cc]: It's part of the overall SQL philosophy. I propose we embrace it fully rather than find reasons to be different.

File: test/cql/lwt_test.result
Comment (avikivity) [reply test/cql/lwt_test.result:2052]: IS NULL does't exist in LWT yet.

File: cql3/column_condition.cc
Comment (avikivity) [reply cql3/column_condition.cc:22]: Will fix.

File: cql3/column_condition.cc
Comment (avikivity) [reply cql3/column_condition.cc:55]: There's no way to modify in-place. It bothers me too, but of minor importance since it's done at prepare time.

File: cql3/statements/modification_statement.cc
Comment (avikivity) [reply cql3/statements/modification_statement.cc]: I'll change it.

File: cql3/column_condition.cc
Comment (nyh) [reply cql3/column_condition.cc:55]: If there is no way to modify in-place, maybe there can at least be a way for a function that does nothing (e.g., if there is no LIKE at all) to at least not copy - maybe just move the old expression?

File: cql3/expr/expression.cc
Comment (nyh) [reply cql3/expr/expression.cc:2660]: But these names "NONE" etc. are new to this series, aren't they? Or maybe I just missed it and it's pre-existing code that you moved?

File: cql3/expr/prepare_expr.cc
Comment (nyh) [reply cql3/expr/prepare_expr.cc:1290]: Ok.

File: cql3/column_condition.cc
Comment (avikivity) [reply cql3/column_condition.cc:55]: Yes, but why? We have to traverse the entire thing, so complexity is unchanged. It's not going to impact performance.

File: cql3/expr/expression.cc
Comment (avikivity) [reply cql3/expr/expression.cc:2660]: These are overrides to an virtual functions defined by an abstract base class. I must define them.

**Comment by avikivity** [comment]
Update:
 
 - reworded commit messages mentioning "SPECIAL NULL", which no longer exists
 - internal function for optimized like matcher now mentions the pattern it matches
 - rearranged modification_statement::applies_to()::condition_applies to
   eliminate an unnecessary check if the row happens to be missing

File: cql3/column_condition.cc
Comment (nyh) [reply cql3/column_condition.cc:55]: Complexity doesn't change but may it slows this step by a large constant factor?
Anyway, I agree this can be optimized later.

Comment (avikivity) [reply cql3/column_condition.cc:55]: These checks will make one case faster (when no changes need to be made) and another case faster (when we have to both check and rewrite). I think prepare should favor simplicity.

File: cql3/expr/prepare_expr.cc
Comment (avikivity) [reply cql3/expr/prepare_expr.cc]: Changed.

**Comment by nyh** [comment]
Thanks. I'll wait for the CI to approve and merge this series, unless someone from the LWT team wants to wait so they can review this series as well? CC @kostja @gleb-cloudius

File: cql3/statements/modification_statement.cc
Comment (avikivity) [reply cql3/statements/modification_statement.cc]: Changed. btw, this code was eliminated later, so the end result is unaffected by this.

**Pull request merged**
